onnxruntime/core/framework/partial_executor.cc

```diff
@@ -0,0 +1,497 @@
+    state_.GetExecutionFrame()->UpdateFetches(fetch_mlvalue_idxs, fetches, session_state.GetInitializedTensors());
+  }
+
+  ExecutionFrame& frame = *(state_.GetExecutionFrame());
+  LOGS(logger, INFO) << "Begin execution";
+  const SequentialExecutionPlan& seq_exec_plan = *session_state.GetExecutionPlan();
+  const auto& exec_plan_vec = seq_exec_plan.execution_plan;
+  VLOGS(logger, 1) << "Size of execution plan vector: " << exec_plan_vec.size();
+
+// Enable TRACE_EXECUTION compile flag to dump execution plan
+#if defined(TRACE_EXECUTION)
+  std::cout << std::make_pair(&seq_exec_plan, &session_state) << std::endl;
+#endif
+
+  const auto& graph_viewer = session_state.GetGraphViewer();
+
+#ifdef CONCURRENCY_VISUALIZER
+  // need unique name for the series. number of nodes should be good enough for a subgraph
+  char series_name[MaxSeriesNameLengthInChars] = "MainGraph";
+  if (graph_viewer->IsSubgraph()) {
+    auto s = graph_viewer->ParentNode()->Name().substr(0, MaxSeriesNameLengthInChars - 1);
+    std::copy(s.cbegin(), s.cend(), series_name);
+  }
+
+  diagnostic::marker_series series(series_name);
+#endif
+
+#ifdef ENABLE_NVTX_PROFILE
+  auto& profile_context = profile::Context::GetInstance();
+  const auto tag = profile_context.GetThreadTagOrDefault(std::this_thread::get_id());
+  profile::NvtxRangeCreator forward_range(
+      "Batch-" + tag + " Forward",
+      profile::Color::White);
+  profile::NvtxRangeCreator backward_range(
+      "Batch-" + tag + " Backward",
+      profile::Color::Black);
+#endif
+
+  for (size_t program_counter = state_.GetProgramCounterStart(); program_counter <= state_.GetProgramCounterEnd(); program_counter += 1) {
+    const auto& node_exec_plan = exec_plan_vec[program_counter];
+    if (terminate_flag_) {
+      LOGS(logger, WARNING) << "Exiting due to terminate flag being set to true.";
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Exiting due to terminate flag being set to true.");
+    }
+
+    auto node_index = node_exec_plan.node_index;
+    const auto& node = *graph_viewer.GetNode(node_exec_plan.node_index);
+
+#ifdef CONCURRENCY_VISUALIZER
+    series.write_flag(node.Name().c_str());
+#endif
+
+#ifdef ENABLE_NVTX_PROFILE
+    if (node.Description() != "Backward pass" && !forward_range.IsBeginCalled()) {
+      // Start timing forward pass when encountering the first forward node.
+      forward_range.Begin();
+    } else if (node.Description() == "Backward pass" && !backward_range.IsBeginCalled() && forward_range.IsBeginCalled()) {
+      // Start timing backward pass when encountering the first backward node.
+      // In the meanwhile, forward range ends.
+      forward_range.End();
+      backward_range.Begin();
+    }
+#endif
+
+    auto p_op_kernel = session_state.GetKernel(node_index);
+    if (p_op_kernel->KernelDef().OpName() == "YieldOp") {
+      // Do not execute YieldOp (it is an no-op anyways).
+      // Decrement the reference count of tensors that are not needed beyond this point.
```

Comment:
What triggers exiting of the forward pass if we just skip over the YieldOp and only_execute_path_to_fetches is always false?

Reply:
We set start and end program counter values in TrainingAgent. Forward pass stops right before yield op, backward pass starts from yield op.